gradle-plugin/src/main/java/com/squareup/hephaestus/plugin/HephaestusPlugin.kt

```diff
@@ -58,101 +78,129 @@ open class HephaestusPlugin : Plugin<Project> {
-              .register(
-                  compileTask.name + "CheckIncrementalCompilationHephaestus",
-                  DisableIncrementalCompilationTask::class.java
-              ) {
-                it.compileTask = compileTask
-                it.configurationName = configurationNameForIncrementalChecks
-              }
+    // Use this signal to share state between DisableIncrementalCompilationTask and the Kotlin
+    // compile task. If the plugin classpath changed, then DisableIncrementalCompilationTask sets
+    // the signal to false.
+    val incrementalSignal = project.gradle.sharedServices
+        .registerIfAbsent("incrementalSignal${project.path}", IncrementalSignal::class.java) { }
+
+    val disableIncrementalCompilationAction: (String) -> Unit = { compileTaskName ->
+      // Disable incremental compilation, if the compiler plugin dependency isn't up-to-date.
+      // This will trigger a full compilation of a module using Hephaestus even though its
+      // source files might not have changed. This workaround is necessary, otherwise
+      // incremental builds are broken. See https://youtrack.jetbrains.com/issue/KT-38570
+      val disableIncrementalCompilationTaskProvider = project.tasks.register(
+          compileTaskName + "CheckIncrementalCompilationHephaestus",
+          DisableIncrementalCompilationTask::class.java
+      ) { task ->
+        task.pluginClasspath.from(
+            project.configurations.getByName(PLUGIN_CLASSPATH_CONFIGURATION_NAME)
+        )
+        task.incrementalSignal.set(incrementalSignal)
+      }
 
-          compileTask.dependsOn(checkIfCompilerPluginChangedTaskProvider)
+      project.tasks.named(compileTaskName, KotlinCompile::class.java) { compileTask ->
+        compileTask.dependsOn(disableIncrementalCompilationTaskProvider)
 
-          compileTask.doFirst {
-            compileTask.logger.info(
-                "Hephaestus: Incremental compilation enabled: ${compileTask.incremental}"
-            )
+        compileTask.doFirst {
+          // If the signal is set, then the plugin classpath changed. Apply the setting that
+          // DisableIncrementalCompilationTask requested.
+          val incremental = incrementalSignal.get().incremental
+          if (incremental != null) {
+            compileTask.incremental = incremental
           }
+
+          compileTask.logger.lifecycle(
+              "Hephaestus: Incremental compilation enabled: ${compileTask.incremental} (compile)"
+          )
         }
+      }
+    }
+
+    if (isAndroidProject) {
+      project.androidVariantsConfigure { variant ->
+        val compileTaskName = "compile${variant.name.capitalize(US)}Kotlin"
+        disableIncrementalCompilationAction(compileTaskName)
+      }
+    } else {
+      // The Java plugin has two Kotlin tasks we care about: compileKotlin and compileTestKotlin.
+      val taskNames = listOf("compileKotlin", "compileTestKotlin")
```

Comment:
Nit: I'm not a fan of the intermediate variables that don't serve a purpose. 

`listOf("compileKotlin", "compileTestKotlin")
.forEach(disableIncrementalCompilationAction)`

Would be cleaner in my opinion. Your call!

Reply:
I prefer the intermediate variables. It makes it more clear that the strings in the list represent task names.

Reply:
Same here in this case.